firestore/collgroupref.go

```diff
@@ -36,3 +46,146 @@ func newCollectionGroupRef(c *Client, dbPath, collectionID string) *CollectionGr
+	for _, part := range qp {
+		queries = append(queries, part.ToQuery())
+	}
+	return queries, nil
+}
+
+// getPartitions returns a slice of QueryPartition objects, describing a start
+// and end range to query a subsection of the collection group. partitionCount
+// must be a positive value and the number of returned partitions may be less
+// than the requested number if providing the desired number would result in
+// partitions with very few documents.
+func (cgr CollectionGroupRef) getPartitions(ctx context.Context, partitionCount int) ([]queryPartition, error) {
+	orderedQuery := cgr.query().OrderBy(DocumentID, Asc)
+
+	if partitionCount <= 0 {
+		return nil, errors.New("a positive partitionCount must be provided")
+	} else if partitionCount == 1 {
+		return []queryPartition{{CollectionGroupQuery: orderedQuery, StartAt: "", EndBefore: ""}}, nil
+	}
+
+	db := cgr.c.path()
+	ctx = withResourceHeader(ctx, db)
+
+	// CollectionGroup Queries need to be ordered by __name__ ASC.
+	query, err := orderedQuery.toProto()
+	if err != nil {
+		return nil, err
+	}
+	structuredQuery := &firestorepb.PartitionQueryRequest_StructuredQuery{
+		StructuredQuery: query,
+	}
+
+	// Uses default PageSize
+	pbr := &firestorepb.PartitionQueryRequest{
+		Parent:         db + "/documents",
+		PartitionCount: int64(partitionCount),
+		QueryType:      structuredQuery,
+	}
+	cursorReferences := make([]*firestorepb.Value, 0, partitionCount)
+	iter := cgr.c.c.PartitionQuery(ctx, pbr)
+	for {
+		cursor, err := iter.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+
+			return nil, fmt.Errorf("GetPartitions: %v", err)
+		}
+		cursorReferences = append(cursorReferences, cursor.Values...)
+	}
+
+	// From Proto documentation:
+	// To obtain a complete result set ordered with respect to the results of the
+	// query supplied to PartitionQuery, the results sets should be merged:
+	// cursor A, cursor B, cursor M, cursor Q, cursor U, cursor W
+	// Once we have exhausted the pages, the cursor values need to be sorted in
+	// lexicographical order by segment (areas between '/').
+	sort.Sort(byFirestoreValue(cursorReferences))
+
+	partitionQueries := make([]queryPartition, 0, len(cursorReferences))
+	previousCursor := ""
+
+	for _, cursor := range cursorReferences {
+		cursorRef := cursor.GetReferenceValue()
+
+		// remove the root path from the reference, as queries take cursors
+		// relative to a collection
+		cursorRef = cursorRef[len(orderedQuery.path)+1:]
+
+		qp := queryPartition{
+			CollectionGroupQuery: orderedQuery,
+			StartAt:              previousCursor,
+			EndBefore:            cursorRef,
+		}
+		partitionQueries = append(partitionQueries, qp)
+		previousCursor = cursorRef
+	}
+
+	// In the case there were no partitions, we still add a single partition to
+	// the result, that covers the complete range.
+	lastPart := queryPartition{
+		CollectionGroupQuery: orderedQuery,
+		StartAt:              "",
+		EndBefore:            "",
```

Comment:
Nit: Again, I think these are unnecessary, but maybe y'all want to be explicit with these offsets.